libyul/backends/evm/StackHelpers.h

```diff
@@ -0,0 +1,408 @@
+	// Returns a positive integer n, if the slot at the given target offset needs n more copies.
+	// Returns a negative integer -n, if the slot at the given target offsets occurs n times too many.
+	// Returns zero if the amount of occurrences, in the current source layout, of the slot at the given target offset
+	// matches the desired amount of occurrences in the target.
+	{ ops.targetMultiplicity(targetOffset) } -> std::convertible_to<int>;
+	// Returns true, iff any slot is compatible with the given target offset.
+	{ ops.targetIsArbitrary(targetOffset) } -> std::convertible_to<bool>;
+	// Returns the number of slots in the source layout.
+	{ ops.sourceSize() } -> std::convertible_to<size_t>;
+	// Returns the number of slots in the target layout.
+	{ ops.targetSize() } -> std::convertible_to<size_t>;
+	// Swaps the top most slot in the source with the slot `depth` slots below the top.
+	// In terms of EVM opcodes this is supposed to be a `SWAP<depth>`.
+	// In terms of vectors this is supposed to be `std::swap(source.at(source.size() - depth - 1, source.top))`.
+	{ ops.swap(depth) };
+	// Pops the top most slot in the source, i.e. the slot at offset ops.sourceSize() - 1.
+	// In terms of EVM opcodes this is `POP`.
+	// In terms of vectors this is `source.pop();`.
+	{ ops.pop() };
+	// Dups or pushes the slot that is supposed to end up at the given target offset.
+	{ ops.pushOrDupTarget(targetOffset) };
+};
+*/
+/// Helper class that can perform shuffling of a source stack layout to a target stack layout via
+/// abstracted shuffle operations.
+template</*ShuffleOperationConcept*/ typename ShuffleOperations>
+class Shuffler
+{
+public:
+	/// Executes the stack shuffling operations. Instantiates an instance of ShuffleOperations
+	/// in each iteration. Each iteration performs exactly one operation that modifies the stack.
+	/// After shuffle, source and target have the same size and all slots in the source layout are
+	/// compatible to the slots at the same target offset.
+	template<typename... Args>
+	static void shuffle(Args&&... args)
+	{
+		bool needsMoreShuffling = true;
+		// The shuffling algorithm should always terminate in polynomial time, but we provide a limit
+		// in case it does not terminate due to a bug.
+		size_t iterationCount = 0;
+		while (iterationCount < 1000 && (needsMoreShuffling = shuffleStep(std::forward<Args>(args)...)))
+			++iterationCount;
+		yulAssert(!needsMoreShuffling, "Could not create stack layout after 1000 iterations.");
+	}
+private:
+	// If dupping an ideal slot causes a slot that will still be required to become unreachable, then dup
+	// the latter slot first.
+	// @returns true, if it performed a dup.
+	static bool dupDeepSlotIfRequired(ShuffleOperations& _ops)
+	{
+		// Check if the stack is large enough for anything to potentially become unreachable.
+		if (_ops.sourceSize() < 15)
+			return false;
+		// Check whether any deep slot might still be needed later.
```

Comment:
Yeah, fair enough, probably not the clearest way to put this.
An example of the situation is:
```
[X 0] => [X 0 0 0 0 0 0 ... 0 0 0 0 0 X]
```
If I had unlimited stack access, then the ideal thing to do is to just keep pushing zeroes, and then in the very end dup the X from the very bottom. But if it's more than 16 zeroes, that won't work.
So after 15 zeroes, I need to bring up a copy of ``X`` already, even if it's not optimal, since otherwise I wouldn't reach it anymore and I still "need it later", i.e. the slot with ``X`` still needs to be touched later on, either with a DUP or a SWAP.